adaptive/src/main/java/com/google/accompanist/adaptive/TwoPane.kt

```diff
@@ -0,0 +1,411 @@
+                        LayoutDirection.Rtl -> splitLeft
+                    }
+                    constraints.copy(minWidth = width, maxWidth = width)
+                } else {
+                    val height = constraints.maxHeight - splitBottom
+                    constraints.copy(minHeight = height, maxHeight = height)
+                }
+            val firstPlaceable = firstMeasurable.measure(constraints.constrain(firstConstraints))
+            val secondPlaceable = secondMeasurable.measure(constraints.constrain(secondConstraints))
+
+            firstPlaceable.placeRelative(0, 0)
+            val detailOffsetX =
+                if (isHorizontalSplit) {
+                    constraints.maxWidth - secondPlaceable.width
+                } else {
+                    0
+                }
+            val detailOffsetY =
+                if (isHorizontalSplit) {
+                    0
+                } else {
+                    constraints.maxHeight - secondPlaceable.height
+                }
+            secondPlaceable.placeRelative(detailOffsetX, detailOffsetY)
+        }
+    }
+}
+
+/**
+ * A strategy for configuring the [TwoPane] component, that is responsible for the meta-data
+ * corresponding to the arrangement of the two panes of the layout.
+ */
+@Stable
+fun interface TwoPaneStrategy {
+    /**
+     * Calculates the split result in local bounds.
+     */
+    fun calculateSplitResult(
+        density: Density,
+        layoutDirection: LayoutDirection,
+        layoutCoordinates: LayoutCoordinates
+    ): SplitResult
+}
+
+/**
+ * Returns the specification for where to place a split in [TwoPane] as a result of
+ * [TwoPaneStrategy.calculateSplitResult]
+ */
+class SplitResult(
+
+    /**
+     * Whether the split is vertical or horizontal
+     */
+    val isHorizontalSplit: Boolean,
+
+    /**
+     * The area that is nether a `start` pane or an `end` pane, but a separation between those two.
+     * In case width or height is 0 - it means that the area itself is a 0 width/height, but the
+     * place within the layout is still defined.
+     */
+    val splitArea: Rect,
+)
+
+/**
+ * Returns a [TwoPaneStrategy] that will place the slots vertically or horizontally if there is a
+ * horizontal or vertical fold respectively.
+ *
+ * If there is no fold, then the [fallbackStrategy] will be used instead.
+ */
+public fun TwoPaneStrategy(
+    fallbackStrategy: TwoPaneStrategy,
+    windowGeometry: WindowGeometry,
+): TwoPaneStrategy = HorizontalTwoPaneStrategy(
+    fallbackStrategy = VerticalTwoPaneStrategy(
+        fallbackStrategy = fallbackStrategy,
+        windowGeometry = windowGeometry,
+    ),
+    windowGeometry = windowGeometry
+)
+
+/**
+ * Returns a [TwoPaneStrategy] that will place the slots horizontally if there is a vertical fold.
+ *
+ * If there is no horizontal fold, then the [fallbackStrategy] will be used instead.
+ */
+public fun HorizontalTwoPaneStrategy(
+    fallbackStrategy: TwoPaneStrategy,
```

Comment:
Agreed, the "falllback" approach is nice from an implementation detail (check for folds first, then fallback to something else), but folding is not going to be the most common result of the strategy.

I think a `CompositionLocal` as mentioned above could help in this department, so that there's less of a lift required to opt-in to fold aware behavior.

Reply:
Let me know what you think of the latest iteration. I removed the "wrapping" behavior in the implementation in favor of a list of conditional strategies that ends with one that is not conditional.

I also removed all public entry points except those that support fold-aware behavior by default for the fraction and offset entry points.

For the more general entry points, I renamed the argument from `fallbackStrategy` to `defaultStrategy`.

I think the new API does make it much easier to have foldable support by default, at the cost of being a bit stranger.